tools/gnur-runtime/src/main/java/org/renjin/gnur/api/Rinternals.java

```diff
@@ -536,15 +536,22 @@ public static int IS_LONG_VEC(SEXP x) {
   }
 
   public static int LEVELS(SEXP x) {
-    throw new UnimplementedGnuApiMethod("LEVELS");
+    final SEXP levels = x.getAttribute(Symbols.LEVELS);
+    if (Null.INSTANCE == levels) {
+      return IntVector.NA;
+    } else {
+      return levels.asInt();
+    }
   }
 
   public static int SETLEVELS(SEXP x, int v) {
-    throw new UnimplementedGnuApiMethod("SETLEVELS");
+    AbstractSEXP abstractSEXP = (AbstractSEXP) x;
+    abstractSEXP.unsafeSetAttributes(x.getAttributes().copy().set(Symbols.LEVELS, IntVector.valueOf(v)));
+    return LEVELS(x);
   }
 
-  public static Object DATAPTR(SEXP x) {
-    if(x instanceof IntVector | x instanceof LogicalVector) {
+  public static Ptr DATAPTR(SEXP x) {
```

Comment:
The change to the return type here will change the signature and break packages that have been previously been compiled against Renjin: they will fail to load with a `NoSuchMethodError`. Unfortunately there's not really any clean way to simply `@Deprecate` this method and a new one with a different return type because of constraints on the Java language. 

If we leave the return type as `java.lang.Object` then our compiler IIRC will just insert an extra cast to Ptr when receiving the return value, which is annoying but not the worst.

Can we leave the return type alone for the moment?